p2p/http/libp2phttp.go

```diff
@@ -0,0 +1,525 @@
+// HTTP semantics with libp2p. Can use a libp2p stream transport or stock HTTP
+// transports. This API is experimental and will likely change soon. Implements libp2p spec #508.
+package libp2phttp
+
+import (
+	"bufio"
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"sync"
+
+	lru "github.com/hashicorp/golang-lru/v2"
+	host "github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/network"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/libp2p/go-libp2p/core/protocol"
+	gostream "github.com/libp2p/go-libp2p/p2p/gostream"
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+const ProtocolIDForMultistreamSelect = "/http/1.1"
+const PeerMetadataLimit = 8 << 10 // 8KB
+const PeerMetadataLRUSize = 256   // How many different peer's metadata to keep in our LRU cache
+
+// TODOs:
+// - integrate with the conn gater and resource manager
+// - skip client auth
+// - Support listenAddr option to accept a multiaddr to listen on (could handle h3 as well)
+// - Support withStreamHost option to accept a streamHost to listen on
+
+type WellKnownProtocolMeta struct {
+	Path string `json:"path"`
+}
+
+type WellKnownProtoMap map[protocol.ID]WellKnownProtocolMeta
+
+// WellKnownHandler is an http.Handler that serves the .well-known/libp2p resource
+type WellKnownHandler struct {
+	wellknownMapMu   sync.Mutex
+	wellKnownMapping WellKnownProtoMap
+}
+
+// StreamHostListen retuns a net.Listener that listens on libp2p streams for HTTP/1.1 messages.
+func StreamHostListen(streamHost host.Host) (net.Listener, error) {
+	return gostream.Listen(streamHost, ProtocolIDForMultistreamSelect)
+}
+
+func (h *WellKnownHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	// Check if the requests accepts JSON
+	accepts := r.Header.Get("Accept")
+	if accepts != "" && !(strings.Contains(accepts, "application/json") || strings.Contains(accepts, "*/*")) {
+		http.Error(w, "Only application/json is supported", http.StatusNotAcceptable)
+	}
+
+	if r.Method != "GET" {
+		http.Error(w, "Only GET requests are supported", http.StatusMethodNotAllowed)
+	}
+
+	// Return a JSON object with the well-known protocols
+	h.wellknownMapMu.Lock()
+	mapping, err := json.Marshal(h.wellKnownMapping)
+	h.wellknownMapMu.Unlock()
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	w.WriteHeader(http.StatusOK)
```

Comment:
[Unnecessary if status is `http.StatusOK`](https://pkg.go.dev/net/http@go1.20.7#ResponseWriter)
```suggestion
```

Reply:
True, although there is no harm in being explicit.